runelite-client/src/main/java/net/runelite/client/plugins/inferno/InfernoNPC.java

```diff
@@ -25,162 +25,472 @@
+			possibleNextLocations.add(currentPosition.dx(dx));
+
+			if (Math.abs(target.getX() - currentPosition.getX()) != Math.abs(target.getY() - currentPosition.getY())
+				|| new WorldArea(currentPosition, npc.getTransformedDefinition().getSize(), npc.getTransformedDefinition().getSize()).distanceTo(target) != 1)
+			{
+				possibleNextLocations.add(currentPosition.dy(dy));
+			}
+
+			WorldPoint bestNextLocation = null;
+			WorldArea bestNexArea = null;
+			for (WorldPoint possibleNextLocation : possibleNextLocations)
+			{
+				//System.out.println("Checking loc: " + possibleNextLocation.getX() + ", " + possibleNextLocation.getY());
+				if (possibleNextLocation.getX() == currentPosition.getX() && possibleNextLocation.getY() == currentPosition.getY())
+				{
+					//System.out.println("Position is the same, continuing");
+					continue;
+				}
+
+				final WorldArea possibleNextArea = new WorldArea(possibleNextLocation, npc.getTransformedDefinition().getSize(),
+					npc.getTransformedDefinition().getSize());
+
+				boolean stillPossible = true;
+				for (WorldPoint obstacle : realObstacles)
+				{
+					if (possibleNextArea.intersectsWith(new WorldArea(obstacle, 1, 1)))
+					{
+						//System.out.println("Intersected with obstacle, discarding position");
+						stillPossible = false;
+						break;
+					}
+				}
+				if (stillPossible)
+				{
+					bestNextLocation = possibleNextLocation;
+					bestNexArea = possibleNextArea;
+					break;
+				}
+			}
+
+			if (bestNextLocation != null && bestNexArea != null)
+			{
+				boolean hasLos = new WorldArea(target, 1, 1).hasLineOfSightTo(client, bestNexArea);
+				boolean hasRange = this.getType().getDefaultAttack() == Attack.MELEE ? bestNexArea.isInMeleeDistance(target)
+					: bestNexArea.distanceTo(target) <= this.getType().getRange();
+
+				if (hasLos && hasRange)
+				{
+					//System.out.println("Move location has LOS, can walk to: true");
+					safeSpotCache.put(bestNextLocation, 1);
+					return true;
+				}
+
+				currentPosition = bestNextLocation;
+			}
+		}
+	}
+
+	boolean couldAttackPrevTick(Client client, WorldPoint lastPlayerLocation)
+	{
+		return new WorldArea(lastPlayerLocation, 1, 1).hasLineOfSightTo(client, this.getNpc().getWorldArea());
+	}
+
+	void gameTick(Client client, WorldPoint lastPlayerLocation, boolean finalPhase)
+	{
+		safeSpotCache.clear();
+
+		if (ticksTillNextAttack > 0)
+		{
+			this.ticksTillNextAttack--;
+		}
+
+		//Jad animation detection
+		if (this.getType() == Type.JAD && this.getNpc().getAnimation() != -1 && this.getNpc().getAnimation() != this.lastAnimation)
+		{
+			final InfernoNPC.Attack currentAttack = InfernoNPC.Attack.attackFromId(this.getNpc().getAnimation());
+
+			if (currentAttack != null && currentAttack != Attack.UNKNOWN)
+			{
+				this.updateNextAttack(currentAttack, this.getType().getTicksAfterAnimation());
+			}
+		}
+
+		if (ticksTillNextAttack <= 0)
+		{
+			if (this.getType() == Type.ZUK)
```

Comment:
A switch statement would work nicely here in my opinion

Reply:
> 
> 
> A switch statement would work nicely here in my opinion

I agree.